### PR TITLE
Add retry policy (exponential backoff)

### PR DESCRIPTION
**Pull request by mark-r-m**
Update retry policy to exponential backoff and update tests

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -16,8 +16,11 @@ repositories {
 }
 
 dependencies {
-	implementation("com.fasterxml.jackson.module:jackson-module-kotlin")
 	implementation("org.springframework.boot:spring-boot-starter-batch")
+	implementation("org.springframework.retry:spring-retry:1.2.4.RELEASE")
```
Comment (benj-fletch) [inline build.gradle.kts]: Can we remove this version and let the dependency resolver sort the version for us?

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-bin.zip
+#distributionUrl=https\://nexus3.ucds.io/repository/thirdparty/org/gradle/distributions/gradle/5.5.1/gradle-5.5.1-bin.zip
```
Comment (benj-fletch) [inline gradle/wrapper/gradle-wrapper.properties:4]: remove commented line

File: src/main/kotlin/app/services/impl/HttpKeyService.kt [hunk 3]
```diff
@@ -21,44 +23,65 @@ import java.net.URLEncoder
 @Service
 class HttpKeyService(private val httpClientProvider: HttpClientProvider) : KeyService {
 
+    companion object {
+        val logger: Logger = LoggerFactory.getLogger(HttpKeyService::class.toString())
+
+        // Will retry at 1s, 2s, 4s, 8s, 16s then give up (after a total of 31 secs)
+        const val maxAttempts = 5
```
Comment (benj-fletch) [inline src/main/kotlin/app/services/impl/HttpKeyService.kt:30]: DW-2680 states a 10 `maxAttempts`, should we stay consistent to that?

File: src/main/kotlin/app/services/impl/HttpKeyService.kt [hunk 4]
```diff
@@ -21,44 +23,65 @@ import java.net.URLEncoder
...
-                            decryptedKeyCache[cacheKey] = dataKeyResult.plaintextDataKey
-                            dataKeyResult.plaintextDataKey
+        try {
+            return if (decryptedKeyCache.containsKey(cacheKey)) {
+                decryptedKeyCache[cacheKey]!!
+            }
+            else {
+                httpClientProvider.client().use { client ->
+                    val dksUrl = "$dataKeyServiceUrl/datakey/actions/decrypt?keyId=${URLEncoder.encode(encryptionKeyId, "US-ASCII")}"
+                    logger.info("dataKeyServiceUrl: '$dksUrl'.")
```
Comment (benj-fletch) [inline src/main/kotlin/app/services/impl/HttpKeyService.kt]: This log will be a bit chatty for not much benefit?

File: src/main/kotlin/app/services/impl/HttpKeyService.kt [hunk 5]
```diff
@@ -21,44 +23,65 @@ import java.net.URLEncoder
...
         }
+        catch (ex: DataKeyDecryptionException) {
+            throw ex
+        }
+        catch (ex: DataKeyServiceUnavailableException) {
+            throw ex
+        }
+        catch (ex: Exception) {
+            throw DataKeyServiceUnavailableException("Error contacting data key service: $ex")
+        }
```
Comment (benj-fletch) [inline src/main/kotlin/app/services/impl/HttpKeyService.kt:83]: A little neater perhaps?
```
when(ex) {
    is DataKeyDecryptionException, is DataKeyServiceUnavailableException -> {
        throw ex
    }
    else -> throw DataKeyServiceUnavailableException("Error contacting data key service: $ex")
}
```

File: gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (mark-r-m) [reply gradle/wrapper/gradle-wrapper.properties:4]: It's a useful record as it's used by us internally.

File: src/main/kotlin/app/services/impl/HttpKeyService.kt (hunk 3 above)
Comment (mark-r-m) [reply src/main/kotlin/app/services/impl/HttpKeyService.kt:30]: No, I'll update the ticket - we looked at a reasonable total time.

File: src/main/kotlin/app/services/impl/HttpKeyService.kt (hunk 4 above)
Comment (mark-r-m) [reply src/main/kotlin/app/services/impl/HttpKeyService.kt]: Not very, no, as we don't call it very often. useful for debugging.

File: src/main/kotlin/app/services/impl/HttpKeyService.kt (hunk 5 above)
Comment (mark-r-m) [reply src/main/kotlin/app/services/impl/HttpKeyService.kt:83]: Cool.

File: build.gradle.kts (hunk 1 above)
Comment (mark-r-m) [reply build.gradle.kts]: Yup that's good.

File: gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (benj-fletch) [reply gradle/wrapper/gradle-wrapper.properties:4]: Fair enough

File: src/main/kotlin/app/services/impl/HttpKeyService.kt (hunk 3 above)
Comment (benj-fletch) [reply src/main/kotlin/app/services/impl/HttpKeyService.kt:30]: Ok!

File: src/main/kotlin/app/services/impl/HttpKeyService.kt (hunk 4 above)
Comment (benj-fletch) [reply src/main/kotlin/app/services/impl/HttpKeyService.kt]: Ok, might be worth downgrading to `.debug` level but if it's not called often then not too bothered either way

**Review by benj-fletch — APPROVED** [review]

**Pull request merged**